pkg/client/client.go

```diff
@@ -0,0 +1,590 @@
+	if err != nil {
+		return nil, err
+	}
+
+	if cfg.RetrievalConfigurer != nil {
+		if err := dataTransfer.RegisterTransportConfigurer(&retrievalmarket.DealProposal{}, cfg.RetrievalConfigurer); err != nil {
+			return nil, err
+		}
+	}
+
+	if err := dataTransfer.Start(ctx); err != nil {
+		return nil, err
+	}
+
+	retrievalEventPublisher := eventpublisher.NewEventPublisher(ctx)
+
+	client := &RetrievalClient{
+		blockstore:              cfg.Blockstore,
+		dataTransfer:            dataTransfer,
+		graphSync:               graphSync,
+		host:                    cfg.Host,
+		payChanMgr:              cfg.PayChannelManager,
+		retrievalEventPublisher: retrievalEventPublisher,
+		transport:               dtTransport,
+	}
+
+	retrievalEventPublisher.Subscribe(client)
+
+	return client, nil
+}
+
+func (rc *RetrievalClient) RetrieveContentFromPeerAsync(
+	ctx context.Context,
+	peerID peer.ID,
+	minerWallet address.Address,
+	proposal *retrievalmarket.DealProposal,
+) (result <-chan retriever.RetrievalResult, onProgress <-chan uint64, gracefulShutdown func()) {
+	gracefulShutdownChan := make(chan struct{}, 1)
+	resultChan := make(chan retriever.RetrievalResult, 1)
+	progressChan := make(chan uint64)
+	internalCtx, internalCancel := context.WithCancel(ctx)
+	go func() {
+		defer internalCancel()
+		result, err := rc.retrieveContentFromPeerWithProgressCallback(internalCtx, peerID, minerWallet, proposal, func(bytes uint64) {
+			select {
+			case <-internalCtx.Done():
+			case progressChan <- bytes:
+			}
+		}, gracefulShutdownChan)
+		resultChan <- retriever.RetrievalResult{result, err}
+	}()
+	return resultChan, progressChan, func() {
+		gracefulShutdownChan <- struct{}{}
+	}
+}
+
+func (rc *RetrievalClient) RetrieveContentFromPeerWithProgressCallback(
+	ctx context.Context,
+	peerID peer.ID,
+	minerWallet address.Address,
+	proposal *retrievalmarket.DealProposal,
+	progressCallback func(bytesReceived uint64),
+) (*retriever.RetrievalStats, error) {
+	return rc.retrieveContentFromPeerWithProgressCallback(ctx, peerID, minerWallet, proposal, progressCallback, nil)
+}
+
+func (rc *RetrievalClient) retrieveContentFromPeerWithProgressCallback(
+	ctx context.Context,
+	peerID peer.ID,
+	minerWallet address.Address,
+	proposal *retrievalmarket.DealProposal,
+	progressCallback func(bytesReceived uint64),
+	gracefulShutdownRequested <-chan struct{},
+) (*retriever.RetrievalStats, error) {
+	if progressCallback == nil {
+		progressCallback = func(bytesReceived uint64) {}
+	}
+
+	log.Infof("Starting retrieval with miner peer ID: %s", peerID)
+
+	ctx, span := tracer.Start(ctx, "rcRetrieveContent")
+	defer span.End()
+
+	// Stats
+	startTime := time.Now()
+	totalPayment := abi.NewTokenAmount(0)
+
+	rootCid := proposal.PayloadCID
+	var chanid datatransfer.ChannelID
+	var chanidLk sync.Mutex
+
+	willingToPay := rc.payChanMgr != nil // The PayChannelManager can be nil if we don't want to pay for retrievals
```

Comment:
@hannahhoward @rvagg Not sure about checking rc.payChanMgr vs cfg.PayChannelManager.

I also don't really like how `payChanAddr` and `payChanLane` end up in the void if the client is not willing to pay or a payment is not required. In the future, I could see having an abstracted `RetrievalPayment` struct that can manage payments for an individual retrieval.